lunabot_embedded/src/teensy_driver_node.cpp

```diff
@@ -16,87 +18,134 @@ extern "C" {
 #include "pb_encode.h"
 }
 
+using namespace std;
+
 #define BUF_SIZE 64
+#define MAX_ANGLE_DELTA_DEG 15
+
+// Left encoder reports half the speed due to hardware error
+#define LEFT_ENCODER_MULTIPLIER 2
+
+#define LEAKY_INTEGRATOR_ALPHA 0.6
 
 uint8_t buf[BUF_SIZE];
 
+RobotState prev_state = RobotState_init_zero;
 RobotState state = RobotState_init_zero;
 RobotEffort effort = RobotEffort_init_zero;
+lunabot_msgs::RobotState prev_state_msg;
+float prev_raw_drive_ang_left = 0;
+float prev_raw_drive_ang_right = 0;
+
+RingBuffer<20> left_buffer;
+RingBuffer<10> right_buffer;
+
+float prev_drive_left_vel = 0;
+float prev_drive_right_vel = 0;
+
+double prev_time = 0;
 
 pb_ostream_t sizestream = {0};
 
 void recv(ros::Publisher &pub) {
-    int status;
-    /* Create a stream that reads from the buffer. */
-    pb_istream_t stream = pb_istream_from_buffer(buf, sizeof(buf));
-    /* Now we are ready to decode the message. */
-    pb_decode(&stream, RobotState_fields, &state);
-    lunabot_msgs::RobotState state_msg;
-    state_msg.act_right_curr = state.act_right_curr;
-    state_msg.drive_right_curr = state.drive_right_curr;
-    state_msg.drive_left_curr = state.drive_left_curr;
-    state_msg.lead_screw_curr = state.lead_screw_curr;
-    state_msg.dep_curr = state.dep_curr;
-    state_msg.exc_curr = state.exc_curr;
-    state_msg.act_ang = state.act_ang;
-    state_msg.drive_right_ang = state.drive_right_ang;
-    state_msg.drive_left_ang = state.drive_left_ang;
-    state_msg.lead_screw_ang = state.lead_screw_ang;
-    state_msg.dep_ang = state.dep_ang;
-    pub.publish(state_msg);
+  int status;
+  /* Create a stream that reads from the buffer. */
+  pb_istream_t stream = pb_istream_from_buffer(buf, sizeof(buf));
+  /* Now we are ready to decode the message. */
+  pb_decode(&stream, RobotState_fields, &state);
+  lunabot_msgs::RobotState state_msg;
+  state_msg.act_right_curr = adc_to_current_ACS711_15A(state.act_right_curr);
+  state_msg.drive_right_curr = adc_to_current_ACS711_15A(state.drive_right_curr);
+  state_msg.drive_left_curr = adc_to_current_ACS711_15A(state.drive_left_curr);
+  state_msg.lead_screw_curr = adc_to_current_ACS711_15A(state.lead_screw_curr);
+  state_msg.dep_curr = adc_to_current_ACS711_15A(state.dep_curr);
+  state_msg.exc_curr = adc_to_current_ACS711_31A(state.exc_curr);
+  state_msg.act_ang = state.act_ang;
+
+  angle_noise_rej_filter(&state.drive_left_ang, &prev_raw_drive_ang_left, MAX_ANGLE_DELTA_DEG);
```

Comment:
Unless I'm misunderstanding this, the filtering logic here is basically set the current angle value equal the previous value if it spiked too much. However, this will set the "velocity" the encoder predicts to be 0, which I don't think should be the behavior. Would it not be better to assume the encoder is at the same velocity? Or to not publish velocity? Also when the next velocity calculation is performed after an invalid value should it not be divided by the total accumulated dt between the last valid state and the current one?

Also how often does this filtering happen? If it happens often I'm a little concerned that we won't be able to get good estimations regardless of filtering and whatnot

Reply:
Your understanding is correct. I wonder if there's a more elegant way to apply this filter. can refactor to have the filter output a bool to if the value was rejected, which can be used to choose to recalculate the velocity. the dt should only include the time between successive measurements, so yes, the dt should be accumulated. Will make those changes

Reply:
I am marking this for a future issue/fix as we got working code without it. Until we write offline unit tests for this, would rather make this change and validate on the real robot to make sure it doesn't break anything.